### PR TITLE
feat: lazy load sparklines

### PR DESCRIPTION
**Pull request by cartcrom**

### CODE REVIEW

File: src/components/Charts/SparklineChart.tsx [hunk 1]
```diff
@@ -51,5 +73,3 @@ function SparklineChart({ width, height, tokenData, pricePercentChange, timePeri
     />
   )
 }
-
-export default React.memo(SparklineChart)
```
Comment (cmcewen) [inline src/components/Charts/SparklineChart.tsx:55]: y no memo?

File: src/components/Tokens/TokenTable/TokenTable.tsx [hunk 2]
```diff
@@ -75,73 +79,51 @@ export function LoadingTokenTable({ rowCount }: { rowCount?: number }) {
...
-        </>
-      )
-    }
+    return (
+      <GridContainer>
+        <HeaderRow />
+        <TokenDataContainer>
+          {tokens.map(
+            (token, index) =>
+              token && (
```
Comment (cmcewen) [inline src/components/Tokens/TokenTable/TokenTable.tsx:115]: where would token be null/undefined here?

File: src/graphql/data/TopTokens.ts [hunk 3]
```diff
@@ -121,240 +128,43 @@ function useFilteredTokens(tokens: PrefetchedTopToken[]) {
...
-  const hasMore = !tokens || tokens.length < prefetchedSelectedTokensWithoutPriceHistory.length
   const environment = useRelayEnvironment()
+  const [sparklines, setSparklines] = useState<SparklineMap>({})
+  useMemo(() => {
+    fetchQuery<TopTokensSparklineQuery>(environment, tokenSparklineQuery, {
+      duration,
+      chain,
+    }).subscribe({
+      next(data) {
+        const map: SparklineMap = {}
```
Comment (cmcewen) [inline src/graphql/data/TopTokens.ts:149]: is there ever a case where the data will be the same and we trigger a re-render without needing to?

File: src/components/Charts/SparklineChart.tsx (hunk 1 above)
Comment (cartcrom) [reply src/components/Charts/SparklineChart.tsx:55]: Didn't mean to remove, can put back in

File: src/components/Tokens/TokenTable/TokenTable.tsx (hunk 2 above)
Comment (cartcrom) [reply src/components/Tokens/TokenTable/TokenTable.tsx:115]: Relay makes everything nullable, and the backend might return null if there's no data available

File: src/graphql/data/TopTokens.ts (hunk 3 above)
Comment (cartcrom) [reply src/graphql/data/TopTokens.ts:149]: The query cache works here just fine, it should only rerequest data if it is stale

**Review by cmcewen — APPROVED** [review]

File: src/components/Charts/SparklineChart.tsx [hunk 4]
```diff
@@ -1,30 +1,53 @@
...
   const theme = useTheme()
   // for sparkline
-  const pricePoints = filterPrices(tokenData?.market?.priceHistory) ?? []
-  const hasData = pricePoints.length !== 0
-  const startingPrice = hasData ? pricePoints[0] : DATA_EMPTY
-  const endingPrice = hasData ? pricePoints[pricePoints.length - 1] : DATA_EMPTY
+  const pricePoints = tokenData?.address ? sparklineMap[tokenData.address] : null
+
+  // Don't display if there's one or less pricepoints
+  if (!pricePoints || pricePoints.length <= 1) {
```
Comment (vm) [inline src/components/Charts/SparklineChart.tsx:41]: should this be an empty state rather than loading state?

**Review by vm — COMMENTED** [review]
should we log the row number backwards when it's descending?

File: src/components/Charts/SparklineChart.tsx (hunk 4 above)
Comment (cmcewen) [reply src/components/Charts/SparklineChart.tsx:41]: the failure case is that there is an error with fetching the data right?

**Comment by awruble** [comment]
yes - I think we should log the row number backwards when it's descending

**Pull request merged**